A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+Admin Interface API
+----
+* Author(s): lidiz
+* Approver: markroth
+* Status: Draft
+* Implemented in: C++, Golang, Java
+* Last updated: 2021-01-27
+* Discussion at: TBD
+
+## Abstract
+
+This proposal describes a convenience API in each gRPC language to improve the usability of creating a gRPC server with admin services.
+
+
+## Background
+
+gRPC Admin Interface is part of the Proxyless Service Mesh Observability project which includes new gRPC admin services and a CLI tool (yet to release). The admin services are responsible for exposing states of the gRPC library via gRPC protocol, and the CLI tool will help users to better interpret the fetched data from the gRPC application. Currently, the existing admin server should at least expose two admin services:
+
+* Channelz: the channel tracing library that collects stats about a channel’s traffic, underlying sockets, and credentials ([proto definition](https://github.com/grpc/grpc/blob/master/src/proto/grpc/channelz/channelz.proto));
+* [Client Status Discovery Service](https://github.com/envoyproxy/envoy/blob/main/api/envoy/service/status/v3/csds.proto): the xDS config dump library that dumps the accepted xDS configuration with config status;
+
+However, the number of admin services might increase in the future. Solutions with no code changes, like env, or config files, are not desired (see rationale section below). On the other hand, as a library, gRPC would like to promote the usage of admin services. So, there is a need for an ideally one LOC to create and start an admin server.
+
+In addition, by providing new APIs, we abstract away the individual services with the concept admin services. We can also achieve **dynamic admin services selection**, for example, we **only add CSDS service if the xDS package is pulled in**. In this way, we won’t force users to take unnecessary dependencies.
+
+
+### Types of Services
+
+To help communicate the scenarios, we can define all services into several types, from the view of administrative functions:
+
+* **Primary service**: the user-provided services that handle the application logic;
+* **Binary-global service**: the services that expose the state of the entire gRPC library, and they serve the same information regardless of which server they have been registered (e.g., Channelz, CSDS);
+* **Per-server service**: the services that distribute the state of other registered services on the same server (e.g., Health, Reflection).
+
+The binary-global services contain sensitive information about traffic, configuration, even security, hence we hope users can set up the services securely in our recommended usage or the upcoming new API.
+
+The recommended way of securing the admin service port is binding it to either `localhost` or using a `UDS` port.
```

Comment:
`localhost` or UDS are not all that secure since anybody with the client can access it without having to authenticate. Is that level of security sufficient (may be it is)?

Reply:
It's the security option with best trade-offs. The TLS is secure, but requires certificate setup which is not that easy for ad-hoc debugging. Since this server might used to debug xDS, it might be a chicken-egg problem if we use xDS mTLS.

As discussed below, the ideal solution is to have a sort-of console to console credential.

Reply:
Agree about "console-to-console" creds (haven't heard this term before) being the best solution in this context.